### ISSUE TITLE
The case for always putting `crate.lock` under version control

### ISSUE THREAD

**Issue by mosteo** [alire-project/alire#468]
This should be shorter than #466 so I hope a regular issue will suffice. @Fabien-Chouteau, I'm explaining my thinking so you can pick it apart. 

TL;DR: I don't find a reason not to always put the lockfile under version control.

**Context**: The lockfile is used to store the current solution, so when it is needed it is not re-resolved. Optionally, it stores pins (to a concrete version and to folders) that override the regular dependency in the manifest.

**The issue**: Up to now I was following the [Cargo reasoning](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html) for not putting the lockfile under version control. There, it is advised that it should only be done for crates that produce a final product: an executable or static library. This makes sense to ensure this final product has a reproducible build. Meanwhile, a crate that is a reusable library should not lock dependencies, to allow more combinations in conjunction with other libraries.

The difference in Alire is that we don't recursively use lockfiles found in dependencies as pins. There is only one lockfile at play, which is that of the root crate. Hence, putting lockfiles under VC for projects intended to be used as dependencies (libraries) has no effect in our case. This corresponds to the Cargo advice of not putting the lockfile under VC.

For projects intended to be used as roots (executables, static libs), we indeed use a preexisting lockfile, so it makes sense to put it under VC. Furthermore, if for some reason the lockfile solution isn't desired, it can be regenerated with `alr update` anyway. 

Now, why this difference? I don't know the reasoning in Cargo, but in our case we recursively collect dependencies during solving, starting from the root crate, using manifests only. It doesn't make sense to reuse partial solutions from dependencies, since version resolution requires to globally know all restrictions on a crate; several partial solutions may be incompatible.

A separate issue is that we _might_ want to honor pins in dependencies lockfiles. Here, the difference is that in Cargo the mere lockfile existence is the pin, whereas we have an explicit _pinned_ attribute, so still we wouldn't hit Cargo's problem even if every crate comes with a lockfile but without pins, which is the usual.

**Conclusion**: we have no reason not to put the lockfile under VC, whereas it is convenient for root crates and may be convenient for other information stored in the lockfile in the future.

**Comment by Fabien-Chouteau** [comment]
> TL;DR: I don't find a reason not to always put the lockfile under version control.

I agree with all the above.
 
> **Conclusion**: we have no reason not to put the lockfile under VC, whereas it is convenient for root crates and may be convenient for other information stored in the lockfile in the future.

A couple things come to mind:

 - ~~Do we need a separate lockfile or can the lock info be included in the manifest?~~ Just as I am writing this I realize that this would be another machine edit of the manifest...
 - Another pro would be that people can mix `Alire` and git submodules during development/prototyping by pinning a submodule dir.
 - As I understand, the current lockfiles containt a duplicate of the crate manifest for the locked crates. Is that necessary? can't we have just the pinned version of the crate?

**Comment by mosteo** [comment]
> * ~Do we need a separate lockfile or can the lock info be included in the manifest?~ Just as I am writing this I realize that this would be another machine edit of the manifest...

I also thought about this. If we end needing a machine section in the manifest to support `alr with`, we could as well store the lockfile information there too. Although I feel some uneasiness that I cannot pinpoint about this. Having a single Alire file would be cleaner for users though.

> * Another pro would be that people can mix `Alire` and git submodules during development/prototyping by pinning a submodule dir.

I'm not sure I 100% follow here. Would pins to remote repositories as we plan to have work for this case as well?

> * As I understand, the current lockfiles contain a duplicate of the crate manifest for the locked crates. Is that necessary? can't we have just the pinned version of the crate?

That's correct. Certainly we could reduce the amount of information we store. With the current solution we reuse the release load/save code and we don't need to query the index for any information about the release. But yes, we could do with just the milestone and origin/integrity hashes (to eventually be able to verify reproducibility).

**Comment by Fabien-Chouteau** [comment]
> > * Another pro would be that people can mix `Alire` and git submodules during development/prototyping by pinning a submodule dir.
> 
> I'm not sure I 100% follow here. Would pins to remote repositories as we plan to have work for this case as well?

Sure that would be a better way to do it :)
Either way, that is another reason to put the lockfile in version control.

**Comment by mosteo** [comment]
Implemented by #505.
